device/pci_device.cpp

```diff
@@ -0,0 +1,431 @@
+/*
+ * SPDX-FileCopyrightText: (c) 2023 Tenstorrent Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <cstdint>
+#include <vector>
+#include <fcntl.h>  // for ::open
+#include <unistd.h> // for ::close
+#include <sys/ioctl.h> // for ioctl
+#include <sys/mman.h>  // for mmap, munmap
+#include <linux/pci.h> // for PCI_SLOT, PCI_FUNC
+
+#include "pci_device.hpp"
+#include "architecture_implementation.h"
+#include "ioctl.h"
+#include "device/tt_arch_types.h"
+#include "device/driver_atomics.h"
+
+#include "common/assert.hpp"
+#include "common/logger.hpp"
+
+int find_device(const uint16_t device_id) {
+    // returns device id if found, otherwise -1
+    const char device_name_pattern [] = "/dev/tenstorrent/%u";
+    char device_name[sizeof(device_name_pattern) + std::numeric_limits<unsigned int>::digits10];
+    std::snprintf(device_name, sizeof(device_name), device_name_pattern, (unsigned int)device_id);
+    int device_fd = ::open(device_name, O_RDWR | O_CLOEXEC);
+    // LOG2 ("find_device() open call returns device_fd: %d for device_name: %s (device_id: %d)\n", device_fd, device_name, device_id);
+    return device_fd;
+}
+
+tt::ARCH detect_arch(uint16_t pcie_device_id, int pcie_revision_id) {
+    if (pcie_device_id == 0xfaca){
+        return tt::ARCH::GRAYSKULL;
+    } else if (pcie_device_id == 0x401e && pcie_revision_id == 0x01){
+        return tt::ARCH::WORMHOLE_B0;
+    } else if (pcie_device_id == 0x401e){
+        TT_THROW("Wormhole is not supported. Please use Wormhole B0 instead.");
+        return tt::ARCH::WORMHOLE;
+    } else if (pcie_device_id == 0xb140){
+        return tt::ARCH::BLACKHOLE;
+    } else {
+        TT_THROW("Unknown pcie device id that does not match any known architecture: ", pcie_device_id);
+    }
+}
+
+// --------------------------------------------------------------------------------------------------------------
+// --------------------------------------------------------------------------------------------------------------
+// --------------------------------------------------------------------------------------------------------------
+
+TTDevice::TTDevice(int device_id, int logical_device_id){
```

Comment:
Just to make sure I understand, you (mostly) copied this class from tt_silicon_driver class?

Reply:
Yes